### PR TITLE
PYIC-5537: Added validation to the journey events submitted from front-end to back-end

### PR DESCRIPTION
**Pull request by RogVanGDS**
### What changed

Added validation to the journey events submitted from front-end to back-end

### Why did it change

If a user disabled Javascript in the browser there is nothing stopping them from submitting multiple journey events to the journey engine. 

Given that most of the journey events that a user can submit are “next” or “end”,  a user can quite easily skip or jump pages, because we transition the user’s state whenever we receive an event (provided that event exists on the current user state retrieved by the backend), regardless of the page the user sees on the screen.

### Issue tracking

- [PYIC-5537](https://govukverify.atlassian.net/browse/PYIC-5537)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

[PYIC-5537]: https://govukverify.atlassian.net/browse/PYIC-5537?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java [hunk 1]
```diff
@@ -153,12 +159,20 @@ public Map<String, Object> handleRequest(Map<String, String> input, Context cont
         } catch (SqsException e) {
             return StepFunctionHelpers.generateErrorOutputMap(
                     HttpStatus.SC_INTERNAL_SERVER_ERROR, ErrorResponse.FAILED_TO_SEND_AUDIT_EVENT);
+        } catch (UnsupportedEncodingException e) {
+            return StepFunctionHelpers.generateErrorOutputMap(
+                    HttpStatus.SC_INTERNAL_SERVER_ERROR, ErrorResponse.valueOf(e.getMessage()));
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (RogVanGDS) [inline lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java]: Change this

File: libs/cri-storing-service/src/test/java/uk/gov/di/ipv/core/library/cristoringservice/CriStoringServiceTest.java [hunk 2]
```diff
@@ -163,7 +163,7 @@ void storeVcsShouldProcessVcsAndSendAuditEvents() throws Exception {
     }
 
     @Test
-    void storeVcsShouldProcessTicfVcsAndSendAuditEvents() throws Exception {
+    void storeTicfVcsShouldProcessVcsAndSendAuditEvents() throws Exception {
```
Comment (RogVanGDS) [inline libs/cri-storing-service/src/test/java/uk/gov/di/ipv/core/library/cristoringservice/CriStoringServiceTest.java]: Revert this change

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java [hunk 3]
```diff
@@ -1,24 +1,31 @@
...
 
 import java.io.IOException;
 import java.util.Map;
 
 public class StateMachine {
     public static final String DELIMITER = "/";
-
     private final Map<String, State> states;
 
+    private static final Logger LOGGER = LogManager.getLogger();
```
Comment (RogVanGDS) [inline lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java]: Revert

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java [hunk 4]
```diff
@@ -27,11 +31,27 @@ public State transition(String startState, String event, JourneyContext journeyC
...
         }
 
         return newState;
     }
+
+    private boolean isPageOrCriStateAndOutOfSync(BasicState basicState, String currentPage) {
+        return basicState.getResponse() instanceof PageStepResponse pageStepResponse
+                        && !pageStepResponse.getPageId().equals(currentPage)
+                || basicState.getResponse() instanceof CriStepResponse criStepResponse
+                        && !criStepResponse.getCriId().equals(currentPage);
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java]: Can `criStepResponse.getCriId().equals(currentPage)` ever be true? (i.e. can we just remove that condition)

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java [hunk 5]
```diff
@@ -27,11 +31,27 @@ public State transition(String startState, String event, JourneyContext journeyC
                     String.format("Unknown state provided to state machine: %s", startState));
         }
 
+        if (currentPage != null && state instanceof BasicState basicState) {
+            if (isPageOrCriStateAndOutOfSync(basicState, currentPage)) {
+                return state;
+            } else if (basicState.getResponse() instanceof ProcessStepResponse) {
+                throw new UnknownStateException(
+                        String.format("Unknown state provided to state machine: %s", startState));
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java]: I think this error message isn't quite right - should it be something like `Unexpected page event received in process state`?

**Review by Joe-Edwards-GDS — COMMENTED** [review]
This is looking good I think

File: lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java (hunk 4 above)
Comment (RogVanGDS) [reply lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java]: Ah yes I forgot we refactored the frontend to remove this.

I guess there's an argument to still validate against this in case we change the front-end logic in the future and it still would be an invalid path, but happy to remove if you think it's not necessary?

Comment (RogVanGDS) [reply lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java]: Actually on second thoughts we probably should remove it to avoid confusion

**Review by Joe-Edwards-GDS — DISMISSED** [review]
This looks good (but you've still got a conflict in secrets :( )

**Review by Joe-Edwards-GDS — APPROVED** [review]
